README.md
# Simple Interpreter
## Description
This is an interpreter for the programming language Simple. It allows you to write and run Simple code, which is a simple and easy to learn programming language.

## Requirements
To use this interpreter, you will need:

Python 3.6 or later
The colorama module, which can be installed using pip install colorama
Usage
To use the interpreter, run the script with the Simple code as an argument:

```bash
python main.py -f path/to/your/code.simp
```
Features
This interpreter supports the following features:

- Printing to the console
- Reading input from the user
- Delaying execution with the sleep command
- Single line comments with //
- Jumping to a specific line of code with the goto command
- Declaring and assigning variables with the var command
- Performing arithmetic operations with variables
- Conditional statements with the if command
- Defining and calling functions with the DEF and RUNDEF commands
- 
## Author
This interpreter was created by ousmblueninja.

# Documentaion


The Simple Interpreter supports the following instructions:

Printing
Use the print command to print a message or the value of a variable to the console:

```
print "Hello, World!";
END;
```
```
var x="10";
print "$x";
END;
```
Input
Use the input command to read input from the user:

```
print "What is your name?";
input in; << saves the input text to the variable in
print "Hello";
print "$in";
print "!";
END;
```

Delaying Execution
Use the sleep command to delay the execution of the program for a specified number of seconds:

```
sleep 2;
END;
```
Single Line Comments

Use the // symbol to add a single line comment to your code:
```
// This is a comment; << must end in ";"
print "Hello, World!";
END;
```
Jumping to a Specific Line
Use the goto command to jump to a specific line of code:

```
goto 10;
END;
```
Declaring and Assigning Variables
Use the var command to declare and assign a value to a variable:

```
var x = 10;
END;
```
You can**t** also perform arithmetic operations with variables, **YET**:

```
var x = 10;
var y = 20;
var z = $x + $y;
END;
```
Conditional Statements
Use the if command to execute a block of code if a condition is met:

```
var x = 10;
if ($x=="10"){11} << goto Line Number 11 if x is equal to 10
if ($x=="10"){*main} << goto function main if x is equal to 10
END;
```
Defining and Calling Functions
Use the DEF and RUNDEF commands to define and call functions:

```
DEF myFunction;
print "This is my function";
ENDDEF;

RUNDEF myFunction;
END;
```

## List of commands

1. print
2. input
3. sleep
4. goto
5. RESET << Clears all variable
6. END
7. var
8. ALOC << alocates a variable name
9. if
10. PSTK << prints all variable
11. DEF
12. DEFEND
13. RUNDEF
14. CLR 

MD/CHANGELOG.md
added var and ALOC 

made var overwrite existing data

added runtime -t

added if statments

changed if statements to allow {*functionName}


main.py
import os, sys,  time
import getopt
import simple

errors = False

global DEBUG
DEBUG = True

start = 0.01 # default start time
end = 0.01

def getArgs(argv):
   global start
   global inputfile
   inputfile = ''
   try:
      opts, args = getopt.getopt(argv,"hi:o:t",["ifile="])
   except getopt.GetoptError:
      print('test.py -t <tells you run time> -i <inputfile>')
      sys.exit(2)
   for opt, arg in opts:
      if opt == '-h':
         print('test.py -t <tells you run time> -i <inputfile>')
         sys.exit()
      elif opt in ("-i", "--ifile"):
         inputfile = arg
      elif opt in ("-t", "--time"):
        start = time.time()
      
   return inputfile

def clear():
    os.system('cls' if os.name == 'nt' else 'clear')

class b:
  HEADER = '\033[95m'
  OKBLUE = '\033[94m'
  OKCYAN = '\033[96m'
  OKGREEN = '\033[92m'
  WARNING = '\033[93m'
  FAIL = '\u001b[31m'
  ENDC = '\033[0m'
  BOLD = '\033[1m'
  UNDERLINE = '\033[4m'
  WHITE = '\u001b[37m'

def error(lineNum, file, error):
  print(f'{b.ENDC}[{b.FAIL}ERROR{b.ENDC}] {error} on line: {lineNum} \n')

class syntax:
  global inputfile
  global errors
  def __init__(self):
    errors = False
    self.terminator = ";"

  def check(self, line, lineNumber):
    if line[(len(line) - 1 )] == ";":
      #print(lineNumber)
      pass
    
    else:
      error(lineNumber, inputfile, "missing ';'")
      errors = True
    
    
    
  
  
syntax = syntax()

def parce(filename):
    with open(filename, 'r') as file:
        data = file.read()
        
    block = []
    blocks = []
    blockNum = 0
    
    for i in range(len(data)):
        if data[i] == ';' :
            block.append(data[i])
            blockNum += 1
            blocks.append(block)
            syntax.check(blocks[(len(blocks)-1)], len(blocks))
            block = []
        elif data[i] == "\n":
            pass
        else:
            block.append(data[i])
    #with open(f'{inputfile}.a.out', 'a') as f:
    #  f.write(str(blocks))
    return blocks

    



def main(file):
    global errors, end
    #print(f'Parceing {file}')
    simple.startFile()
    parcedFile = parce(file)
    if errors:
      #print("error")
      return
    else:
      #print("not")
      pass

    #time.sleep(10)
    #print(f'Running {file} with {len(parcedFile)} lines of code')
    #time.sleep(0.5)
    clear()
    i = 0
    fileLen = len(parcedFile)
    while i <= fileLen:
        try:
          blockLen = len(parcedFile[i])
        except IndexError:
          error((i+1), file, "unexpected end of file")
          return
        #print(f'{parcedFile[i]} {blockLen}')
        #print(parcedFile[i][blockLen-1])
        if not parcedFile[i][blockLen-1]==";":
          error((i+1), file, "missing ; ")
          return
        
        #print("".join(parcedFile[i]))
        OPT = 0
        OPT = simple.interpriter(parcedFile[i], i)
        #print(OPT)
        if OPT == None:
          error(0, None, "unknown operator")
          return 
        
        #print(i)
        #print(type(i), type(fileLen), type(OPT)) 
        #print(OPT)
        if OPT == "valErr":
           error(0, None, "Sytax Error")
           return
        elif OPT == "END":
          return
        elif OPT == "END1": # error return value
          return
        
          
        if OPT != 0:
           
           if int(OPT) <= 0:
              return
           elif int(OPT) >= 0:
             if int(OPT) > fileLen:
               error(i, None, "Out of range")
               if DEBUG:
                 print(f'[{b.FAIL}Out of range{b.ENDC}] Proram tring to jump to invalid line {i}')
               return
             i = (int(OPT)-1)
             
        else:
          i = i + 1
        
            
              
    
    end = time.time()



    
    
if __name__ == '__main__':
    main(getArgs(sys.argv[1:]))
    if start != 0.01:
      delta = time.time() - start
      #delta = delta/1000000000
    
      print(f'finished in [{round(abs(delta), 0)}] seconds')

UtilHandler.py
import json, time, os, sys


class b:
  HEADER = '\033[95m'
  OKBLUE = '\033[94m'
  OKCYAN = '\033[96m'
  OKGREEN = '\033[92m'
  WARNING = '\033[93m'
  FAIL = '\u001b[31m'
  ENDC = '\033[0m'
  BOLD = '\033[1m'
  UNDERLINE = '\033[4m'
  WHITE = '\u001b[37m'

def error(lineNum, error):
  print(f'{b.ENDC}[{b.FAIL}ERROR{b.ENDC}] {error} on line: {lineNum} \n')


class file:
  def read(self,filename):
    try:
      with open(filename, 'r') as f:
        return f.read()
    except Exception as e:
      return e
  def write(self,filename, data):
    try:
      with open(filename, 'w') as f:
        f.write(data)
    except Exception as e:
      return e
  def append(self,filename, data):
    with open(filename, 'a') as f:
      f.write(data)
  def remove(self,filename):
    try:
      os.remove(filename)
    except Exception as e:
      return e

file = file()




def FileHandle(code):
    opt = code.split('file.')[1]
    #print(opt)
    if opt.startswith('read'):
      filename = opt.split('"')
      saveVar = filename[2].replace('as', '')
      saveVar = saveVar.replace(' ', '')
      saveVar = saveVar.replace(';', '')
      #print(filename, saveVar, filename[1])
      try:
        data = file.read(filename[1])
        #print(file.read(filename[1]))
      except:
        return 0
      #print(saveVar)
      return data, saveVar
    return 0


def exicute(code, lineNum):
  
  if code.startswith('file.'):
    return FileHandle(code)
  else:
    error((lineNum+1), "Unknown Operator")
    exit()
    

simple.py
# interpriter for the programing language " Simple "
# by https://github.com/ousmblueninja
import os, random


LOGFILE = f"./logs/log_{random.randrange(11111,99999)}.log"
print(LOGFILE)

try:
  os.mkdir("./logs")
except FileExistsError as e:
  pass

class b:
  HEADER = '\033[95m'
  OKBLUE = '\033[94m'
  OKCYAN = '\033[96m'
  OKGREEN = '\033[92m'
  WARNING = '\033[93m'
  FAIL = '\u001b[31m'
  ENDC = '\033[0m'
  BOLD = '\033[1m'
  UNDERLINE = '\033[4m'
  WHITE = '\u001b[37m'


def error(lineNum, error):
  print(f'{b.ENDC}[{b.FAIL}ERROR{b.ENDC}] {error} on line: {lineNum} \n')
  with open(LOGFILE, "a") as f:

    DebugData = f"[ERROR] {error} on line: {lineNum}"
    f.write(str(DebugData))
    f.write("\n")


import time

commands = [
  "print", "input", "sleep", "//", "goto", "RESET", "END", "var", "ALOC", "if",
  "PSTK", "DEF", "DEFEND", "RUNDEF", "CLR", "IF"
]

# OLD \/
# [pointer, name, data]
# NEW \/
# [name, data]
data = []

defRun = False
callLn = 0

# [function name, line number]
functions = []

inFunc = False

DEBUG = True


def startFile():
  with open('debug.log', 'w') as f:
    f.write("LineNum  Operator  |  Variables  |  Functions")
    f.write("\n")


def debugger(ln, operator):
  global inFunc, callLn, data, functions, DEBUG, defRun
  with open(LOGFILE, "a") as f:
    spaces = (12 - len(operator))

    DebugData = f"{ln}        {operator}{' '*spaces}|   {data}   |      {functions}"
    f.write(str(DebugData))
    f.write("\n")


def saveToVar(indata, varname):
  global data
  savedVariable = [len(data), varname, indata]

  for i in range(len(data)):
    #print(data[i])
    try:
      if data[i][1] == varname:
        pos = data[i][0]
        data.pop(i)
        savedVariable = [pos, varname, indata]
    except IndexError:
      pass

  data.append(savedVariable)


def interpriter(raw, lineNum):
  ###############################
  #         DEFEND;             #
  ###############################

  # Moved To Top

  global data, inFunc, functions, callLn, defRun, DEBUG
  #print(inFunc)
  # if DEBUG:
  #   saveDebufData(data)
  # print(data, lineNum)
  global commands
  #print(data)
  code = "".join(raw)
  operator = "".join(raw).split(" ")[0].split(";")[0]
  debugger(lineNum, operator)

  if code.startswith(commands[6]):
    code = code.replace(commands[6] + '', "")
    code = code.replace(';', "")
    #print(code)
    if inFunc == False and defRun:
      with open(LOGFILE, "a") as f:

        f.write(str("EOF Program over."))
        f.write("\n")
      return "END"
  elif code.startswith(commands[12]):
    code = code.replace(commands[12], "")
    code = code.replace(';', "")
    inFunc = False
    if defRun == True:
      return callLn
    else:
      return 0

  if inFunc is True:
    return 0

  #print(data)
  #print("".join(raw))
  if code.startswith(commands[0]):
    #print("printing ")
    code = code.replace(commands[0] + ' "', "")
    code = code.replace('";', "")
    if code.startswith("$"):
      code = code.replace("$", "")
      for i in range(len(data)):
        #print(data[i])
        if data[i][1] == code:

          print(data[i][2], end="")

          return 0
      error((lineNum + 1), "Undifined Variable")
      return "END1"

    else:
      for i in range(len(code)):
        if code[i] == "\\" and code[i + 1] == "n":
          print("")
        elif code[i] == "n" and code[i - 1] == "\\":
          print("")
        else:
          print(code[i], end="")

      return 0
  elif code.startswith(commands[1]):
    code = code.replace(commands[1] + ' ', "")
    code = code.replace(';', "")
    with open(LOGFILE, "a") as f:
      f.write(str("(INPUT: taking User Input \n"))
    inData = input("")
    with open(LOGFILE, "a") as f:
      f.write(
        str(
          f'(DEBUG: User input saved as [ {code} ] with data [ {inData} ] \n'))
    savedVariable = [len(data), code, inData]

    for i in range(len(data)):
      #print(data[i])
      try:
        if data[i][1] == code:
          pos = data[i][0]
          data.pop(i)
          savedVariable = [pos, code, inData]
      except IndexError:
        pass

    data.append(savedVariable)

    #print(data)

    return 0
  elif code.startswith(commands[2]):
    code = code.replace(commands[2] + '(', "")
    code = code.replace(');', "")
    #print(code)
    try:
      time.sleep(int(code))
      return 0
    except ValueError:
      return "valErr"

  elif code.startswith(commands[3]):
    return 0
  elif code.startswith(commands[4]):
    code = code.replace(commands[4] + ' ', "")
    code = code.replace(';', "")
    return code
  elif code.startswith(commands[5]):
    data = []
    return 0

  elif code.startswith(commands[7]):
    code = code.replace(commands[7] + ' ', "")
    code = code.replace(';', "")
    #print(code)
    raw = code.split("=")
    #print(raw)
    raw[1] = raw[1].replace('"', "")
    #print(raw)
    if raw[1].startswith('$'):
      #print(raw[1])
      raw[1] = raw[1].replace('$', "")
      #print(raw)
      valid = False
      for i in range(len(data)):
        if data[i][1] == raw[1]:
          raw[1] = data[i][2]
          #print(raw)
          valid = True
          break
        else:
          valid = False
      if not valid:
        error((lineNum + 1), "undeclaired variable")

      pass

    #print("\n\n", code, data, raw)
    inData = raw[1]
    savedVariable = [len(data), raw[0], inData]

    # checks if dta variable exists \/ \/ \/

    for i in range(len(data)):
      #print(data[i])
      try:
        if data[i][1] == raw[0]:
          data.pop(i)
      except IndexError:
        pass

    # ____________________________________________ end

    data.append(savedVariable)

    #print(data)

    return 0
  elif code.startswith(commands[8]):
    code = code.replace(commands[8] + ' ', "")
    code = code.replace(';', "")
    #print("\n", code)
    for i in range(len(data)):
      #print(data[i])
      try:
        if data[i][1] == code:
          data.pop(i)
          return 0
      except IndexError:
        pass
    return 0
  elif code.startswith(commands[9]):
    code = code.replace(commands[9] + ' (', "")
    code = code.replace(';', "")
    #print("\n", code)
    raw = code.split("==")
    raw.append("")
    #print(raw)
    rawTwo = raw[1].split(')')
    rawTwo[0] = rawTwo[0].replace('"', "")
    #print(rawTwo)
    #print(raw, rawTwo, len(rawTwo))
    for i in range(2):
      raw[(i + 1)] = rawTwo[i - 1]

    cash = raw[1]
    raw[1] = raw[2]
    raw[2] = cash
    del cash
    #raw[1] = raw[1].replace("")
    #print(raw)
    if raw[0].startswith("$"):
      raw[0] = raw[0].replace("$", "")
      for i in range(len(data)):

        #print(data[i])
        if data[i][1] == raw[0]:
          #print(data[i][2])
          raw[0] = data[i][2]
    if raw[1].startswith("$"):
      raw[1] = raw[1].replace("$", "")
      for i in range(len(data)):

        #print(data[i])
        if data[i][1] == raw[1]:
          #print(data[i][2])
          raw[1] = data[i][2]
    raw[2] = raw[2].replace("{", "")
    raw[2] = raw[2].replace("}", "")

    #print(raw)

    if raw[2][0] == "*":
      FuncName = raw[2].strip("*")
      FuncLine = 0

      for x, y in functions:
        if x == FuncName:
          FuncLine = y

      raw[2] = FuncLine

      #print(raw, lineNum)

      for i in range(len(functions)):
        if functions[i][0] == FuncName:
          callLn = (lineNum + 1)
          defRun = True
          inFunc = False
          #print(callLn)
        #print(f"goin to ln {functions[i][1]}")
        #print(functions[i][1])
    else:
      FuncLine = raw[2]

    #print(code, raw, "\n")

    if raw[0] == raw[1]:

      #print(raw)
      return (int(FuncLine))
    else:
      return 0

    #for i in range(len(data)):
    #        #print(data[i])
    #        try:
    #          if data[i][1] == code:
    #            pass
    #        except IndexError:
    #          pass

    return 0

  elif code.startswith(commands[10]):
    code = code.replace(commands[10] + ' (', "")
    code = code.replace(';', "")
    print(data)
    return 0

  elif code.startswith(commands[14]):
    # clear terminal
    os.system("clear")

    return 0
    ###############################
    #         DEF;                #
    ###############################
  elif code.startswith(commands[11]):
    code = code.replace(commands[11] + " ", "")
    code = code.replace(';', "")
    #print(code)
    NewFunc = [code, (lineNum + 2)]
    for i in range(len(functions)):
      if functions[i][0] == code:
        error((lineNum + 1), "Function already exists")
        return "END1"
    functions.append(NewFunc)
    #print(functions)
    inFunc = True

    return 0

    ###############################
    #         RUNDEF;             #
    ###############################
  elif code.startswith(commands[13]):
    code = code.replace(commands[13] + " ", "")
    code = code.replace(';', "")
    #print(code)
    if defRun == True:
      defRun = False
      return 0
    for i in range(len(functions)):
      if functions[i][0] == code:
        callLn = (lineNum + 1)
        defRun = True
        inFunc = False
        #print(f"goin to ln {functions[i][1]}")
        return functions[i][1]

    return "valErr"
